include/aos/iam/modules/pkcs11.hpp

```diff
@@ -0,0 +1,237 @@
+
+    /**
+     * Removes certificate chain using top level certificate URL and password.
+     *
+     * @param certURL top level certificate URL.
+     * @param password owner password.
+     * @return Error.
+     */
+    Error RemoveCert(const String& certURL, const String& password) override;
+
+    /**
+     * Removes private key from a module.
+     *
+     * @param keyURL private key URL.
+     * @param password owner password.
+     * @return Error.
+     */
+    Error RemoveKey(const String& keyURL, const String& password) override;
+
+    /**
+     * Returns valid/invalid certificates.
+     *
+     * @param[out] invalidCerts invalid certificate URLs.
+     * @param[out] invalidKeys invalid key URLs.
+     * @param[out] validCerts information about valid certificates.
+     * @return Error.
+     */
+    Error ValidateCertificates(Array<StaticString<cURLLen>>& invalidCerts, Array<StaticString<cURLLen>>& invalidKeys,
+        Array<CertInfo>& validCerts) override;
+
+private:
+    static constexpr auto cEnvLoginType = "CKTEEC_LOGIN_TYPE";
+    static constexpr auto cDefaultTokenLabel = "aos";
+    static constexpr auto cTeeLoginTypeLen = AOS_CONFIG_CERTHANDLER_TEE_LOGIN_TYPE_NAME_LEN;
+    static constexpr auto cSlotListSize = AOS_CONFIG_CERTHANDLER_SLOT_LIST_SIZE;
+    static constexpr auto cUUIDStringLen = AOS_CONFIG_UUID_STR_LEN;
+    static constexpr auto cRSAKeyLength = 2048;
+    static constexpr auto cECSDACurveID = pkcs11::EllipticCurve::eP384;
+    static constexpr auto cPKCS11Scheme = "pkcs11";
+
+    static constexpr auto cLoginTypePublic = "public";
+    static constexpr auto cLoginTypeUser = "user";
+    static constexpr auto cLoginTypeGroup = "group";
+
+    // TEE Client UUID name space identifier (UUIDv4) from linux kernel
+    // https://github.com/OP-TEE/optee_os/pull/4222
+    // use as constant.
+    uuid::UUID cTeeClientUUIDNs;
+
+    using DERCert = StaticArray<uint8_t, crypto::cDERCertSize>;
+
+    struct PendingKey {
+        uuid::UUID                    mUUID;
+        SharedPtr<pkcs11::PrivateKey> mKey;
+    };
+
+    struct SearchObject;
+
+    String                       GetTokenLabel() const;
+    RetWithError<pkcs11::SlotID> GetSlotID();
+    RetWithError<bool>           IsOwned() const;
+
+    Error PrintInfo(pkcs11::SlotID slotId) const;
+
+    Error GetTeeUserPIN(const String& loginType, uint32_t uid, uint32_t gid, String& userPIN);
+    Error GeneratePIN(const String& loginType, const uuid::UUID& space, uint32_t data, String& userPIN);
+    Error GetUserPin(String& pin) const;
+
+    RetWithError<UniquePtr<pkcs11::SessionContext>> CreateSession(bool userLogin, const String& pin);
+
+    Error FindObject(pkcs11::SessionContext& session, const SearchObject& filter, Array<SearchObject>& dst);
+
+    Error TokenMemInfo() const;
+
+    static bool CheckCertificate(const crypto::x509::Certificate& cert, const crypto::PrivateKey& key);
```

Comment:
Why do you need it static?

Reply:
Basically that's only to say that method doesnt require context of an object.
But I will remove static to make prototype uniform.